directdemod/georeferencer.py

```diff
@@ -0,0 +1,215 @@
+
+        self.tle_file = tle_file
+
+    def georef(self, descriptor, output_file, desc=False):
+
+        '''Main georeferencing routine
+
+        Args:
+            descriptor (:obj:`dict`): descriptor dictionary
+            output_file (:obj:`string`): name of the output file
+            desc (:obj:`bool`): descriptor flag
+        '''
+
+        file_name = descriptor["image_name"]
+        image     = mimg.imread(file_name)
+
+        gcps = self.compute_gcps(descriptor, image)
+
+        options = gdal.TranslateOptions(format="GTiff",
+                                        outputSRS=constants.DEFAULT_RS,
+                                        GCPs=gcps)
+
+        gdal.Translate(destName=constants.TEMP_TIFF_FILE,
+                        srcDS=file_name,
+                        options=options)
+
+        options = gdal.WarpOptions(srcSRS=constants.DEFAULT_RS,
+                                    dstSRS=constants.DEFAULT_RS,
+                                    tps=True,
+                                    resampleAlg=GRA_NearestNeighbour)
+
+        gdal.Warp(destNameOrDestDS=output_file,
+                    srcDSOrSrcDSTab=constants.TEMP_TIFF_FILE,
+                    options=options)
+
+        os.remove(constants.TEMP_TIFF_FILE)
+
+        if desc:
+            self.create_desc(descriptor, output_file)
+
+    def georef_os(self, descriptor, output_file, desc=False):
+
+        '''Main georeferencing routine
+
+        Args:
+            descriptor (:obj:`dict`): descriptor dictionary
+            output_file (:obj:`string`): name of the output file
+            desc (:obj:`bool`): descriptor flag
+        '''
+
+        file_name = descriptor["image_name"]
+        image     = mimg.imread(file_name)
+
+        gcps = self.compute_gcps(descriptor, image)
+
+        translate_flags = "-of GTiff -a_srs EPSG:4326"
+        warp_flags = "-r near -tps -s_srs EPSG:4326 -t_srs EPSG:4326"
+
+        translate_query = 'gdal_translate ' + translate_flags + ' ' + self.to_string_gcps(gcps) + ' "' + file_name + '" ' + ' "' + constants.TEMP_TIFF_FILE + '"'
+        warp_query = 'gdalwarp ' + warp_flags + ' "' + constants.TEMP_TIFF_FILE + '" ' + ' "' + output_file + '"'
+
+        os.system(translate_query)
+        os.system(warp_query)
+
+        os.remove(constants.TEMP_TIFF_FILE)
+
+        if desc:
+            self.create_desc(descriptor, output_file)
+
+    def to_string_gcps(self, gcps):
+
+        '''Create string representation of gcp points
+
+        Args:
+            gcps (:obj:`list`): list of gcp points
+        '''
+
+        return " ".join([("-gcp " + str(gcp.GCPPixel) + " " + str(gcp.GCPLine) + " " + str(gcp.GCPX) + " " + str(gcp.GCPY)) for gcp in gcps])
+
+
+    def create_desc(self, descriptor, output_file):
+
+        '''Create descriptor file
+
+        Args:
+            descriptor (:obj:`dict`): descriptor dictionary
+            output_file (:obj:`string`): name of the output file
+        '''
+
+        desc = {
```

Comment:
hmmm.. this descriptor is still needed?
We could use an image format which will store all this info within itself. Did you take a look at geotiff? is it possible to store the needed info within the image. Actually, the only info that the image needs is: each pixel should be associated with a lat/long and none of these other things are required. We have to change the NOAA decoder part which produces the image such that when it writes the image it also create this image which has every pixel tagged with a latlong. By doing this, we can make our code independent of which sat is producing the image. If I want to use a meteor image, that can also produce an image with tagged pixels and I can just use that image also. I suspect further that even there will be variation within the 3 different NOAA sats.

Reply:
Here as you may have notice the generation of descriptor is optional, and I think not needed. 
After georeferencing actually all info is stored in the tiff file. So descriptor is only needed before georeferencing. Just added for the case, but it may be deleted.
I think that if decoders produce descriptor files, it would be more flexible, because maybe not all users would like to georeference the image etc; in addition descriptor can save all the info about the image, including date (tiff doesn't store it), satellite name and so on. So here we just create descriptors and then people could do whatever they want, without any limitations. And there won't be differences between the decoders results (i. e. they will be independent, I described that in the proposal).